test/test_smem_reuse.cpp

```diff
@@ -143,35 +144,247 @@ TEST_F(SmemReuseTest, NeedsReorderedPush) {
   auto fusion = std::make_unique<Fusion>();
   FusionGuard fg(fusion.get());
 
-  int64_t H_int = 5, W_int = 6;
-  auto H = IrBuilder::create<Val>(H_int);
-  auto W = IrBuilder::create<Val>(W_int);
+  int64_t H = 5;
 
-  auto tv0 = full({H}, fusion->oneVal(), DataType::Float);
-  auto tv1 = set(tv0); // pos = a. A = tv1
+  auto tv0 = full(
+      {IrBuilder::create<Val>(H)},
+      fusion->oneVal(),
+      DataType::Float); // pos = a. A = tv0
+  tv0->setMemoryType(MemoryType::Shared);
+
+  auto tv1 =
+      pad(tv0, {fusion->zeroVal(), fusion->oneVal()}); // pos = b. B = tv1
+  tv1->setMemoryType(MemoryType::Shared);
+
+  auto tv2 = mul(tv1, tv1); // pos = c
+
+  auto tv3 = sum(tv2, {0}); // gap between b and c. Can parallelize to sync
+
+  auto tv4 = broadcast(tv3, {true});
+  auto tv5 = mul(tv4, tv1); // pos = d. C = tv5
+  tv5->setMemoryType(MemoryType::Shared);
+
+  auto tv6 = add(tv1, tv1); // pos = e
+  fusion->addOutput(tv6);
+
+  auto tv7 = neg(tv5); // pos = f
+  fusion->addOutput(tv7);
+
+  { // This should not re-use memory
```

Comment:
It seems there's something missing here. This fusion is not parallelized at all, so why does it need a syncthreads to reuse the memory?

Reply:
Even when parallelized, no syncthreads should be necessary for fusions like:

```
__shared__ float X[N];
...
auto t0 = X[threadIdx.x]; // last read of X
...
X[threadIdx.x] = t1; // reuse X without syncthreads
...
```

Reply:
That case would be an inner alias and is supported. Outer alias, when there are separate loops for the two allocations, is not supported and requires sync. We could potentially handle that case without syncing too but it might require more machinery for proving indices are equivalent.

Reply:
In this test all three smem allocations have different size, so they cannot be aliased.

Reply:
They don't need to be aliased. I think this is more about the stack-based reuse logic. Since no tensor is parallelized, we should be able to freely pop allocated tensors without a sync.

Reply:
Even without parallelization I think a war hazard exists since two threads may be writing different tensors to the same location though.

Reply:
Well, without parallelization, by definition, there must be one thread.

Reply:
Oh sorry I thought you meant this tensor is not parallelized. If no tensors are parallelized you are right that we wouldn't need any syncs. Do we need to handle that case?

Reply:
Completely serial cases would be fine to ignore as they are just a synthetic example I came up with.

But how about cases like this?

```
__shared__ float X[blockDim.x * 4];
...
for (int i = 0; i < 4; ++i) {
   auto t0 = X[threadIdx.x + blockDim.x * i]; // last read of X
}
...
for (int i = 0; i < 4; ++i) {
   X[threadIdx.x + blockDim.x * i] = t1; // reuse X without syncthreads
}
```

I actually don't remember all the details and differences of the inner and outer sharing, but isn't this case be outer sharing? And if so, no reuse is allowed without a sync, right? 

I think what's missing here is probably something like what we do for the RAW sync insertion. We use the CA maps to analyze if a read after a write requires a sync. See for example: https://github.com/NVIDIA/Fuser/blob/main/csrc/device_lower/analysis/sync_information.cpp#L234

It's quite complicated and also it's one of those that we would be able to simplify a lot with the new ID graph. So, I think it's fine to leave this as a limitation for now.

Reply:
Ah I see. Yes in this case there are non-overlapping sets of elements written/read across threads, so it's safe. It's definitely very similar analysis to what's done in `SyncMap`; I had originally thought we might just augment `SyncMap` to insert these re-use syncs even. Sounds good on leaving it for now and revisiting it after ID graphs are complete.

Reply:
Can you file it as a TODO issue?

Reply:
Drafted an issue here: https://github.com/NVIDIA/Fuser/issues/769